pkg/providers/amifamily/ami.go

```diff
@@ -126,163 +82,131 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
+	if err != nil {
+		return nil, err
+	}
+	if images, ok := p.cache.Get(fmt.Sprintf("%d", hash)); ok {
+		// Ensure what's returned from this function is a deep-copy of AMIs so alterations
+		// to the data don't affect the original
+		return append(AMIs{}, images.(AMIs)...), nil
 	}
-	return res
+	images := map[uint64]AMI{}
+	for _, query := range queries {
+		if err = p.ec2api.DescribeImagesPagesWithContext(ctx, query.DescribeImagesInput(), func(page *ec2.DescribeImagesOutput, _ bool) bool {
+			for _, image := range page.Images {
+				arch, ok := v1beta1.AWSToKubeArchitectures[lo.FromPtr(image.Architecture)]
+				if !ok {
+					continue
+				}
+				requirementSets := func() []scheduling.Requirements {
+					if knownRequirements, ok := query.KnownRequirements[lo.FromPtr(image.ImageId)]; ok {
+						return lo.Map(knownRequirements, func(r scheduling.Requirements, _ int) scheduling.Requirements {
+							r.Add(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, arch))
+							return r
+						})
+					}
+					return []scheduling.Requirements{scheduling.NewRequirements(scheduling.NewRequirement(v1.LabelArchStable, v1.NodeSelectorOpIn, arch))}
+				}()
+				for _, reqs := range requirementSets {
+					reqsHash := lo.Must(hashstructure.Hash(reqs.NodeSelectorRequirements(), hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true}))
+					// If the proposed image is newer, store it so that we can return it
+					if v, ok := images[reqsHash]; ok {
+						candidateCreationTime, _ := time.Parse(time.RFC3339, lo.FromPtr(image.CreationDate))
+						existingCreationTime, _ := time.Parse(time.RFC3339, v.CreationDate)
+						if existingCreationTime == candidateCreationTime && lo.FromPtr(image.Name) < v.Name {
+							continue
+						}
+						if candidateCreationTime.Unix() < existingCreationTime.Unix() {
+							continue
+						}
+					}
+					images[reqsHash] = AMI{
+						Name:         lo.FromPtr(image.Name),
+						AmiID:        lo.FromPtr(image.ImageId),
+						CreationDate: lo.FromPtr(image.CreationDate),
+						Requirements: reqs,
+					}
+				}
+			}
+			return true
+		}); err != nil {
+			return nil, fmt.Errorf("describing images, %w", err)
+		}
+	}
+	p.cache.SetDefault(fmt.Sprintf("%d", hash), AMIs(lo.Values(images)))
+	return lo.Values(images), nil
 }
 
-func (p *DefaultProvider) getRequirementsFromImage(ec2Image *ec2.Image) scheduling.Requirements {
-	requirements := scheduling.NewRequirements()
-	// Always add the architecture of an image as a requirement, irrespective of what's specified in EC2 tags.
-	architecture := *ec2Image.Architecture
-	if value, ok := v1beta1.AWSToKubeArchitectures[architecture]; ok {
-		architecture = value
+// MapToInstanceTypes returns a map of AMIIDs that are the most recent on creationDate to compatible instancetypes
+func MapToInstanceTypes(instanceTypes []*cloudprovider.InstanceType, amis []v1beta1.AMI) map[string][]*cloudprovider.InstanceType {
```

Comment:
Thinking about this function, I think we only use this for Drift.  Wondering it would be better if we just returned the requirements of the image, and then compared the instance type directly to those requirements as a function of the instance type.
```
	mappedAMIs := amifamily.MapToInstanceTypes([]*cloudprovider.InstanceType{nodeInstanceType}, nodeClass.Status.AMIs)
	if !lo.Contains(lo.Keys(mappedAMIs), instance.ImageID) {
		return AMIDrift, nil
	}
	return "", nil```

Reply:
I think that makes sense, I'll take a look 👍

Reply:
Ok, so we also use this in the resolver when generating launch templates and we're really just piggybacking off it for drift. I think it's fine to leave as is.